client/src/main/scala/skuber/apiextensions/v1/CustomResourceDefinition.scala

```diff
@@ -0,0 +1,145 @@
+package skuber.apiextensions.v1
+
+import play.api.libs.json._
+import play.api.libs.functional.syntax._
+import skuber.json.format.{enumFormatMethods, maybeEmptyFormatMethods, objectMetaFormat}
+import skuber.ResourceSpecification.{Schema, StatusSubresource}
+import skuber.{ListResource, NonCoreResourceSpecification, ObjectEditor, ObjectMeta, ObjectResource, ResourceDefinition, ResourceSpecification, TypeMeta}
+
+final case class CustomResourceDefinition(
+  kind: String = "CustomResourceDefinition",
+  override val apiVersion: String = "apiextensions.k8s.io/v1",
+  metadata: ObjectMeta,
+  spec: CustomResourceDefinition.Spec
+) extends ObjectResource
+
+object CustomResourceDefinition {
+
+  type Spec = NonCoreResourceSpecification
+  val Spec: NonCoreResourceSpecification.type = NonCoreResourceSpecification
+
+  val Scope: ResourceSpecification.Scope.type = ResourceSpecification.Scope
+
+  type Names = ResourceSpecification.Names
+  val Names: ResourceSpecification.Names.type = ResourceSpecification.Names
+
+  type Version = ResourceSpecification.Version
+
+  type Subresources = ResourceSpecification.Subresources
+  type ScaleSubresource = ResourceSpecification.ScaleSubresource
+  type StatusSubresource = ResourceSpecification.StatusSubresource
+
+  type CustomResourceDefinitionList = ListResource[CustomResourceDefinition]
+
+  val crdNames = Names("customresourcedefinitions", "customresourcedefinition", "CustomResourceDefinition", List("crd"))
+
+  val specification =
+    NonCoreResourceSpecification(apiGroup = "apiextensions.k8s.io", version = "v1", scope = Scope.Cluster, names = crdNames)
+
+  def apply(name: String, kind: String): CustomResourceDefinition =
+    CustomResourceDefinition(name, kind, "v1", Scope.Namespaced, None, Nil, Nil)
+
+  def apply(name: String, kind: String, scope: Scope.Value): CustomResourceDefinition =
+    CustomResourceDefinition(name, kind, "v1", scope, None, Nil, Nil)
+
+  def apply(name: String, kind: String, shortNames: List[String]): CustomResourceDefinition =
+    CustomResourceDefinition(name, kind, "v1", Scope.Namespaced, None, shortNames, Nil)
+
+  def apply(name: String, kind: String, scope: Scope.Value, shortNames: List[String], versions: List[Version]): CustomResourceDefinition =
+    CustomResourceDefinition(name, kind, "v1", scope, None, shortNames, versions)
+
+  def apply(
+    name: String,
+    kind: String,
+    version: String,
+    scope: Scope.Value,
+    singular: Option[String],
+    shortNames: List[String],
+    versions: List[Version]
+  ): CustomResourceDefinition = {
+    if (versions.exists(v => v.schema.isEmpty))
+      throw new Exception("Schema must be specified in CRD v1 version")
```

Comment:
so why schema is not mandatory?

Reply:
Ah because the `ResourceSpecification.Version` class is used by both `v1beta1` and `v1` CRDs, but the Schema is only used in `v1` (where it is mandatory). I agree it's a little clunky, but I wasn't sure of the best way around it

Maybe splitting it into a base `Version` trait and then specific classes like `CRDv1Version` (which would include the Schema as a mandatory field) would be better, I'm not sure. Happy with whatever approach you think is best!